rfd/0159-password-status.md

```diff
@@ -0,0 +1,128 @@
+  PASSWORD_STATE_UNKNOWN = 0;
+  // Password is known to be not configured.
+  PASSWORD_STATE_UNSET = 1;
+  // Password is known to be configured.
+  PASSWORD_STATE_SET = 2;
+}
+```
+
+This field will be stored under its own resource key, since there is no top-level local auth secrets object in the database; the password hash, MFA and WebAuthn keys are stored separately. Also, the hash is a plain string, not a structured object, so we can't extend it with another field. The resource key will have the following form: `/web/users/<username>/pwd_state`. The resource will be removed along with other user data upon user deletion, since it is located within the username scope.
+
+`PASSWORD_STATE_UNKNOWN` is deliberately set to 0, since this is the default value when there's no information about the password state in the database.
+
+The state changes to `PASSWORD_STATE_UNSET` whenever another user (or cluster onboarding flow) creates a password reset token for a given user, or removes the password (see [Removing a Password](#removing-a-password)). These are scenarios when the system becomes confident that the user has no password. As a consequence, users created after this change is made will never have their password in an unknown state.
+
+The state changes to `PASSWORD_STATE_SET` whenever user sets/resets their password or successfully signs in using a password. The last scenario, in particular, will allow us to gradually fill in information about existing users that have passwords.
+
+> [!WARNING]
+> The PasswordState flag should _not_ be used for authentication itself. Attempting to shortcut password verification may open a timing attack vulnerability.
+
+### Returning Password State to the Account Settings UI
+
+The old account settings page had separate paths for managing passwords and MFA devices, so there is no "global" endpoint that would return information about all authentication methods. There is only an endpoint that returns a list of MFA devices.
+
+We therefore need to expose a special `/webapi/user/passwordstatus` endpoint that will return the `PasswordStatus` flag value — and only this.
+
+## Removing a Password
+
+The point of the changes described so far is to enable the user to escape a state of their account where it's not possible to enter a state with a password configured. However, it is still true that after the user _sets_ the password, again, it's not explicitly possible (without another user's help) to enter the state where no password is set (short of, perhaps, using a strong password and then discarding it). As a follow-up feature, we should consider providing a way to remove a password. This scenario should work the same way as removing any other authentication method: first, verify user's identity, either by a passwordless key, MFA, or the password itself, then overwrite the password with a randomly generated one, and finally, set the state flag to `PASSWORD_STATE_UNSET`.
```

Comment:
I'm inclined to leave this out of scope for now. What do others think?

Reply:
I don't mind. I consider it a far-fetched proposal. :)